### PR TITLE
doc: Add ADR 003-Implement-with-Rust

### PR DESCRIPTION
**Pull request by roberth**
Why Rust?

### CODE REVIEW

File: doc/architecture/003-Implement-with-Rust.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+- (+) Haskell has strong types
+- (+) Haskell has good performance
+- (+) Haskell has good Nix bindings (`hercules-ci-cnix-expr`), with possibly no changes needed to them
+  - (-) if well written
+- (-) Haskell has a steep learning curve
+- (-) Haskell has a its own garbage collector, making integration of the Nix language potentially riskier
+  - (+) This has proven to be a non-issue in `hercules-ci-agent`
+- (-) Haskell has a smaller community than Rust, which could make it harder to find contributors
+- (~) Haskell can be slow to iterate on, unless most iteration can happen through unit tests, which may be tricky for NixOps, of which a large part is integration testing.
+- (-) The Haskell community broadly does not have a good testing culture. (Praise to those who do)
```
Comment (srid) [inline doc/architecture/003-Implement-with-Rust.md:55]: @Norfairking

Comment (NorfairKing) [reply doc/architecture/003-Implement-with-Rust.md:55]: @srid This isn't an argument against Haskell, this is an argument against working with certain people who write Haskell :P

Comment (NorfairKing) [reply doc/architecture/003-Implement-with-Rust.md:55]: In fact I tried starting that project in the past: https://github.com/NorfairKing/declops

Comment (roberth) [inline doc/architecture/003-Implement-with-Rust.md:55]:
```suggestion
```

Maybe not perpetuate this then :+1:

Comment (roberth) [reply doc/architecture/003-Implement-with-Rust.md]: Removed in separate thread.

> Maybe not perpetuate this then 👍

- https://github.com/nixops4/nixops4/pull/5#discussion_r1587281012

File: doc/architecture/003-Implement-with-Rust.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+### C++
+
+- (+) C++ has good performance.
+- (+) C++ is somewhat easy to package. Local packaging would be the same as Nixpkgs.
+- (-) C++ ecosystem in general still doesn't have a good packaging story yet.
+- (-) C++ does not have Nix bindings.
+  - Nix is implemented in C++, but that C++ is not a stable interface. It still needs C++ bindings based on the C interface, which do not exist, and would seem unappealing.
+- (-) C++ has worse memory safety.
+- (-) C++ has worse types.
```
Comment (roberth) [inline doc/architecture/003-Implement-with-Rust.md:64]:
```suggestion
- (-) C++ type system is trickier (has implicit coercions, template generics aren't types that abstract away internals, resulting in complex error messages)
```

File: doc/architecture/003-Implement-with-Rust.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+## Decision
+
+NixOps 4 will be implemented in Rust.
+
+- (+) General purpose language suitable for the purpose ("systems programming")
+- (+) Memory safety
+- (+) No garbage collector, makes integrating with a garbage-collected language easier
+- (+) Good performance
+- (+) Good Nix packaging, with multiple integrations available, which align with local development method
+- (+) Nix bindings available, although they will need to be improved, but they are mostly de-risked
```
Comment (Radvendii) [inline doc/architecture/003-Implement-with-Rust.md:17]: They existed but we're not using them...

File: doc/architecture/003-Implement-with-Rust.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+- (+) Rust has precedent in the Nix community
+  - nixpkgs-fmt, alejandra
+  - rnix-lsp, nil
+  - tvix
+  - nixpkgs-check-by-name
+  - Nix itself has experimented with a gradual Rust migration, although this was reverted because the gradual migration process was not viable
+- (-) Somewhat steep learning curve
+  - Less than Haskell's
+  - Partially mitigated by AI assistants
```
Comment (Radvendii) [inline doc/architecture/003-Implement-with-Rust.md:26]:
```suggestion
```

File: doc/architecture/003-Implement-with-Rust.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+- (?) JS community is large, but perhaps underrepresented in the Nix community
+
+# Java
+
+- (-) Java startup is too slow for CLI tools
+- (-) Java packaging in Nix is sub-par
+
+# C#
+
+- (-) Impopular in open source community
```
Comment (Radvendii) [inline doc/architecture/003-Implement-with-Rust.md:100]:
```suggestion
- (-) Unpopular in open source community
```

File: doc/architecture/003-Implement-with-Rust.md (hunk 4 above)
Comment (roberth) [inline doc/architecture/003-Implement-with-Rust.md:26]:
```suggestion
```

Not a differentiator anymore

File: doc/architecture/003-Implement-with-Rust.md (hunk 3 above)
Comment (roberth) [reply doc/architecture/003-Implement-with-Rust.md:17]: Hmm, I might not have been consistent about this, removing the AI line. Anyway, decision has been made so let's move on.

**Pull request merged**